display version on settings page

PartyLich opened the pull request:
This PR adds a version number display to the Settings component.

As suggested in discord. Didn't spot a corresponding issue.

RichInSQL commented:
This is awesome, Is there a way to get the Actual Server version to display here too? Is that possible?

TomAFrench commented:
> This is awesome, Is there a way to get the Actual Server version to display here too? Is that possible?

It's doable. The server exposes a `/info` endpoint which we can query the version on. We'd just need to make a query to pull that and then display it. It's less trivial than getting the version of the frontend though.

PartyLich commented:
If the endpoint already exists then it'll be a quick add next time im in front of an editor. 

What I could use help on is the best way to visually present it. I do not profess to be a designer.

⁣--


Sent from a mobile device; please excuse my brevity and typos.​

On Jul 1, 2022, 18:47, at 18:47, Tom French ***@***.***> wrote:
>> This is awesome, Is there a way to get the Actual Server version to
>display here too? Is that possible?
>
>It's doable. The server exposes a `/info` endpoint which we can query
>the version on. We'd just need to make a query to pull that and then
>display it. It's less trivial than getting the version of the frontend
>though.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/actualbudget/actual/pull/112#issuecomment-1172787185
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

jlongster commented on packages/desktop-client/src/components/Settings.js:
```diff
@@ -473,6 +473,42 @@ function SettingsLink({ to, name, style, first, last }) {
+    if (!url || url.indexOf('not-configured') !== -1) return;
+
+    try {
+      const res = await fetch(url + '/info');
```
Generally, the frontend shouldn't ever hit the server directly. It should hit a backend method that talks to the server. Just makes things less confusing when it comes to mixing non-auth and auth methods on the server, etc. You couldn't do this if this endpoint needed to be authenticated, and it's nice to keep the workflow consistent

I'm not 100% sure about making this `/info` endpoint something the app relies but, but I guess it's OK for now. It's provided by a node-specific library and we haven't really thought through the shape of the API. We can probably change it later though.

Can you instead hit it via the backend, and expose a new backend method (it can be called `get-server-version` and live beside `get-server-url`)?

jlongster commented:
Can you post a screenshot of what it looks like @PartyLich? Also isn't this _only_ displaying the server version, not the app version?

PartyLich commented:
> Can you post a screenshot of what it looks like @PartyLich? Also isn't this _only_ displaying the server version, not the app version?

Currently in `client | server` form, but there's no indication (label, mouseover, etc) of what's what.
![image](https://user-images.githubusercontent.com/2792750/178305303-7469e14c-db45-4f52-9bfa-0025f4d7f75b.png)

Without a server, the current placeholder is 'N/A'
![image](https://user-images.githubusercontent.com/2792750/178306602-de04ecb4-f000-4a69-8a5b-af7f2aa6b448.png)

PartyLich commented on packages/desktop-client/src/components/Settings.js:
```diff
@@ -473,6 +473,37 @@ function SettingsLink({ to, name, style, first, last }) {
+        styles.smallText
+      ]}
+    >
+      v{window.Actual.ACTUAL_VERSION} | {version ? `v${version}` : 'N/A'}
```
any issues I might be unaware of with using the `window.Actual` object for the app version? There is a `get-version` handler that could be used instead.

jlongster replied:
(Sorry for the delay)

Nope! This should be correct; I think the `get-version` is left around from older code when the frontend code didn't have direct access to it

PartyLich commented:
Played with adding tooltips via [react-aria's `useTooltipTrigger`](https://react-spectrum.adobe.com/react-aria/useTooltipTrigger.html). Would add an additional dependency, but iirc react-aria is part of the suggested FE direction anyway. It hasn't been pushed to this branch, but there should be gifs below for consideration.

No server:
![actual versions with tooltip](https://user-images.githubusercontent.com/2792750/179057650-de8981fc-e68f-451c-ad5c-f1d06a17d4cd.gif)
With server:
![actual server versions with tooltip](https://user-images.githubusercontent.com/2792750/179057681-16839cab-fdda-4f25-a8ec-dc40b6e42cd2.gif)

There's also a [ready made `Tooltip` component in React Spectrum](https://react-spectrum.adobe.com/react-spectrum/Tooltip.html), which may be preferable.

jlongster commented on packages/loot-core/src/server/main.js:
```diff
@@ -1496,6 +1496,24 @@ handlers['subscribe-sign-out'] = async function() {
 
+handlers['get-server-version'] = async function() {
+  if (!getServer() || getServer().BASE_SERVER === UNCONFIGURED_SERVER) {
+    return { error: 'no-server' };
```
A desire, sure :) Low priority though. In general we need to think through how errors are handled from backend methods.

(I think just `error` is fine here... until we have a strict pattern and figure out how errors should be, it's ok)

jlongster commented:
I like `react-aria` and have mentioned that it's possible we should use it, but haven't decided yet (there are others, but it's likely the right one). I'd rather not introduce it yet just for this small feature.

I think this could use some design love; it's confusing to show two versions without any labels so we should just label them directly instead of hiding that information behind a tooltip. In fact, we should probably include this info as well in the bottom-right corner of the manage files screen (look in `ManagementApp.js`).

I'd say the format can be: `App: v1.0.0 | Server: v1.0.0`. The tabs need to stay centered at the top, so I'd put this information below the tabs centered. Do that and add it to `ManagementApp.js` too and I think we'd be good!

PartyLich commented:
- adjusted formatting (centered below). 
- moved server version fetch to a hook for reuse. Seems reasonable to me, may seem like unnecessary indirection to other readers
- added server version to `ManagementApp`

![actual versions](https://user-images.githubusercontent.com/2792750/179879688-8b8f9c3a-9f36-4592-83c3-45083a68e1e2.gif)

I'd also like to update the link on `ManagementApp` while I'm touching that file. There's a related issue actualbudget/actual-server#17 open. My thought is to point to the repo tag, but that's also a dead link at present [(https://github.com/actualbudget/actual/releases/tag/v4.0.2)](https://github.com/actualbudget/actual/releases/tag/v4.0.2). 
If we're going to tag and release here, it would eventually be the right url pattern, but I don't know that this is the *actual* plan. Thoughts?

jlongster commented:
Awesome thanks!

Yes we should fix that. We need to work on how to manage release notes -- ideally that's what it'd link to. It's also a little weird now that there are two versions because we'll probably have two different releases or change logs for the app and the server, so the should probably be separate buttons.

I'd say let's get this in and follow-up with fixing the rest. For now, we could just remove the button. Just let it be text that isn't clickable.

jlongster commented:
Thank you! Will merge this in and we can iterate :)

SerdaOzun commented:
This was never actually working, right? I've build actual-server mutliple times, both on my server and local, and there isn't a version number displayed in the settings.

PartyLich commented:
> This was never actually working, right? I've build actual-server multiple times (weeks ago and now), both on my server and local, and there isn't a version number displayed in the settings.

If you're building `actual-server`, it's using the latest published npm package for the web client. The commit is absolutely working, it's just *not in an published release*.
